server/service/src/sync/file_synchroniser.rs

```diff
@@ -0,0 +1,221 @@
+};
+
+use super::settings::SyncSettings;
+
+pub static MAX_UPLOAD_ATTEMPTS: i32 = 7 * 24; // 7 days * 24 hours Retry sending for up to for 1 week before giving up
+pub static RETRY_DELAY_MINUTES: i64 = 15; // Doubles each retry until MAX_RETRY_DELAY_MINUTES
+pub static MAX_RETRY_DELAY_MINUTES: i64 = 60; // 1 hour
+
+#[derive(Debug)]
+pub(crate) enum FileSyncError {
+    DatabaseError(RepositoryError),
+    CantFindFile(String),
+    StdIoError(std::io::Error),
+    ReqwestError(reqwest::Error),
+    UploadError(String), //TODO improve error handling (e.g. if it needs to wait to sync etc, maybe handle permanent vs temporary errors?)
+}
+
+impl From<RepositoryError> for FileSyncError {
+    fn from(error: RepositoryError) -> Self {
+        FileSyncError::DatabaseError(error)
+    }
+}
+
+pub struct FileSynchroniser {
+    settings: SyncSettings,
+    service_provider: Arc<ServiceProvider>,
+    static_file_service: Arc<StaticFileService>,
+    client: reqwest::Client,
+}
+
+impl FileSynchroniser {
+    pub(crate) fn new(
+        settings: SyncSettings,
+        service_provider: Arc<ServiceProvider>,
+        static_file_service: Arc<StaticFileService>,
+    ) -> Self {
+        Self {
+            settings,
+            service_provider,
+            static_file_service,
+            client: reqwest::Client::new(),
+        }
+    }
+
+    pub(crate) async fn sync(&self) -> Result<usize, FileSyncError> {
+        let ctx = self.service_provider.basic_context()?;
+
+        // Find any files that need to be uploaded
+        // Pick a file to upload
+        // Upload a file (In future this could be a chunk of data, instead of a whole file)
+        // Update the file record with the progress
+        // Yield to the runtime to check if we've received a pause signal
+
+        // Get any files that need to be sent to central server
+        let sync_file_repo = SyncFileReferenceRowRepository::new(&ctx.connection);
+        let files = sync_file_repo.find_all_to_upload()?;
+
+        // Pick a random file and upload it
+        // A random file is chosen in case one particular file is having problems, others should be able to continue
+        let file = files.choose(&mut rand::thread_rng());
+        match file {
+            Some(file) => {
+                // update the database to say we're uploading the file
+                sync_file_repo.update_status(&SyncFileReferenceRow {
+                    status: SyncFileStatus::Uploading,
+                    ..file.clone()
+                })?;
+
+                let result = self.try_uploading_file(file).await;
+
+                let bytes_uploaded = match result {
+                    Ok(bytes_uploaded) => bytes_uploaded,
+                    Err(err) => {
+                        log::error!("Error uploading file: {:#?}", err);
+
+                        // Update database to record the chunk has failed to upload
+                        if file.retries >= MAX_UPLOAD_ATTEMPTS {
```

Comment:
I'm on the fence about if we want a max upload attempts or not... Feels like it if it's genuinely never going to work we should give up at somepoint but when?

Was thinking if we have a nice UI an admin could trigger a re-try upload, maybe when on a better network connection or something?